### DISCUSSION TITLE
Not working with pyright

### DISCUSSION THREAD

**Discussion by ashwinprasadme** [PurpleMyst/sansio-lsp-client#51] [Q&A] [answered]
I'm trying to get this working with pyright.

I'm running the tests from the repo and added pyright start command, i.e., `"pyright": ["pyright-langserver", "--stdio"]`

The initialization is getting passed, i.e., `tserver.wait_for_message_of_type(lsp.Initialized)`

However, after that, I do not seem to receive any more responses from the server, 

Diagnostic response is not received, even the hover methods do not seem to return anything. 

`diagnostics = tserver.wait_for_message_of_type(lsp.PublishDiagnostics)`

Since I'm new to this, I thought someone can help me here. Please let me know, thank you!

**Comment by Akuli** [comment]
I can look into this too, maybe in a few weeks, because I too would like to use pyright with sansio-lsp-client :)

Things I would try next:
- Try to get pyright to print some information to stderr (stdout is already used for emitting langserver events). Does it have a "verbose" flag or something?
- Print everything that the tests are sending and receiving, not as langserver events but as bytes.
- Try to get pyright to work with an editor or another lsp client library. Again, have a look at the sent and received data.

**Reply by ashwinprasadme** [reply]
Thank you for your response. I shall look into this as well. I will let you know how it goes!

**Comment by PurpleMyst** [comment]
I'm unable to test this directly, however [this discussion](https://github.com/microsoft/pyright/discussions/1635) seems to attest that `rootUri` (which the tests use) is not supported by `pyright`.

It might be that we need to switch to using `workspaceFolders` in the tests to achieve full compatibility, as advised in the discussion I've linked.

**Reply by Akuli** [reply]
I'm new to `workspaceFolders`.
- Is it possible to provide both, so that no matter which one a langserver supports, it will work?
- If it is, should it be done in sansio-lsp-client or in other projects that use sansio-lsp-client?

Also, nice to see that you're still active :)

**Reply by PurpleMyst** [reply]
As it seems, the tests actually _do_ set `workspaceFolders`, however it still seems to not work. I'll see if I can spin up something to test this first-hand.

I'm still alive and kicking! I've just been busy with university + being sick the past few weeks, thought I'd check out my notifications and stuff. 

Also, the Raspberry Pi I run my ZNC bouncer on seems to have spontaneously died and I haven't had the time to try and fix it :\

**Reply by Akuli** [reply]
I now have a linode server that runs [a game I made](https://github.com/Akuli/catris). We could set up a bouncer there if you want.

**Comment by PurpleMyst** [comment]
Here's something: I've noticed `pyright-langserver` sends a [`client/registerCapability` request for `workspace/didChangeWorkspaceFolders`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#workspace_didChangeWorkspaceFolders). The test infrastructure does nothing to handle this and so pyright itself may just be endlessly waiting for our reply to that request, leading it to not publish diagnostics.

**Reply by PurpleMyst** [reply]
Replying to the notification still seems to do nothing

**Comment by PurpleMyst** [comment, marked as answer]
This seems to be related to sending `workspaceFolders` in `initialize`. If you set that property to `None`, the tests seem to run.

`pyright` [creates a default workspace](https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/workspaceMap.ts#L82-L114) for files that you haven't provided a workspace for.

I'll need to look further into how `pyright` handles workspaces and the like, but the fact that it works if you just don't send any `workspaceFolders` in initialization is a start.

**Reply by ashwinprasadme** [reply]
Thanks! I just tested it and it works now, I'm able to get responses from the server as expected.
